bindings/java/iota-wallet-java/lib/native/src/lib.rs

```diff
@@ -65,7 +89,67 @@ pub extern "system" fn Java_org_iota_api_NativeApi_sendMessage(
         .new_string(serde_json::to_string(&response).unwrap())
         .expect("Couldn't create java string!");
 
-    output.into_inner()
+    output.into_raw()
+}
+/// # Safety
+///
+/// Object::from_raw(callback) makes a java object from the callback.
+/// This will crash if callback is java null
+#[no_mangle]
+pub unsafe extern "system" fn Java_org_iota_api_NativeApi_listen(
+    env: JNIEnv,
+    _class: jclass,
+    events: jobjectArray,
+    callback: jobject,
+) -> jstring {
+    let string_count = env.get_array_length(events).expect("Couldn't get array length!");
+    let mut events_list: Vec<WalletEventType> = Vec::with_capacity(string_count.try_into().unwrap());
+
+    for i in 0..string_count {
+        let java_str: String = env
+            .get_string(env.get_object_array_element(events, i).unwrap().into())
+            .expect("Couldn't get java string!")
+            .into();
+        events_list.push(WalletEventType::try_from(&*java_str).expect("Unknon WalletEventType"));
+    }
+
+    // Allocate callback globaly to not lose the reference
+    let global_obj = env
+        .new_global_ref(JObject::from_raw(callback))
+        .expect("Failed to make Global ref");
+
+    let guard = MESSAGE_HANDLER.lock().unwrap();
+    crate::block_on(guard.as_ref().unwrap().listen(events_list, move |e| {
+        let ev_ser = serde_json::to_string(&e).expect("Failed to serialise event");
+
+        // Grab env from VM
+        let vm_guard = VM.lock().unwrap();
+        let env = {
+            let vm = vm_guard.as_ref().unwrap();
+            let mut env = vm.get_env();
+            if env.is_err() {
+                env = vm.attach_current_thread().map(|e| *e);
+            }
+            env.expect("failed to get env")
+        };
+
+        let event = JValue::Object(*env.new_string(ev_ser).expect("Couldn't create event as string!")).to_jni();
+        let cb = JValue::Object(global_obj.as_obj()).to_jni();
+
+        env.call_static_method_unchecked(
+            "org/iota/api/NativeApi",
+            METHOD_CACHE.lock().unwrap().unwrap(),
+            ReturnType::Object,
+            &[event, cb],
+        )
+        .expect("Failed to call handleCallback");
+    }));
+
+    let output = env
```

Comment:
do we even need to return something here?

Reply:
I imagine we want to add better error handing than .expect() as that crashes the vm. SO i left it like this for now so we can reuse the nativeApi.handleClientResponse